proposal-templates/0000-swift-template.md

```diff
@@ -47,107 +75,144 @@ reasonably implement the feature.
 
 ## Source compatibility
 
-Relative to the Swift 3 evolution process, the source compatibility
-requirements for Swift 4 are *much* more stringent: we should only
-break source compatibility if the Swift 3 constructs were actively
-harmful in some way, the volume of affected Swift 3 code is relatively
-small, and we can provide source compatibility (in Swift 3
-compatibility mode) and migration.
-
-Will existing correct Swift 3 or Swift 4 applications stop compiling
-due to this change? Will applications still compile but produce
-different behavior than they used to? If "yes" to either of these, is
-it possible for the Swift 4 compiler to accept the old syntax in its
-Swift 3 compatibility mode? Is it possible to automatically migrate
-from the old syntax to the new syntax? Can Swift applications be
-written in a common subset that works both with Swift 3 and Swift 4 to
-aid in migration?
-
-## Effect on ABI stability
-
-The ABI comprises all aspects of how code is generated for the
+Describe the impact of this proposal on source compatibility.  As a
+general rule, all else being equal, Swift code that worked in previous
+releases of the tools should work in new releases.  That means both that
+it should continue to build and that it should continue to behave
+dynamically the same as it did before.  Changes that cannot satisfy
+this must be opt-in, generally by requiring a new language mode.
+
+This is not an absolute guarantee, and the Language Workgroup will
+consider intentional compatibility breaks if their negative impact
+can be shown to be small and the current behavior is causing
+substantial problems in practice.
+
+For proposals that affect parsing, consider whether existing valid
+code might parse differently under the proposal.  Does the proposal
+reserve new keywords that can no longer be used as identifiers?
+
+For proposals that affect type checking, consider whether existing valid
+code might type-check differently under the proposal.  Does it add new
+conversions that might make more overload candidates viable?  Does it
+change how names are looked up in existing code?  Does it make
+type-checking more expensive in ways that might run into implementation
+limits more often?
+
+For proposals that affect the standard library, consider the impact on
+existing clients.  If clients provide a similar API, will type-checking
+find the right one?  If the feature overloads an existing API, is it
+problematic that existing users of that API might start resolving to
+the new API?
+
+## ABI compatibility
+
+Describe the impact on ABI compatibility.  As a general rule, the ABI
+of existing code must not change between tools releases or language
+modes.  This rule does not apply as often as source compatibility, but
+it is much stricter, and the Language Workgroup generally cannot allow
+exceptions.
```

Comment:
I suppose the discomfort I'm feeling here is that the "ABI compatibility" is a hard constraint and so all of the questions here seem a bit more appropriate as _implementation_ guidance rather than stuff that should really be discussed in the proposal

Reply:
Much like leaving the API resilience section around, we want some assurance that the proposal author has thought about it.  Expanded the wording here with some suggestions and guidance.